runtime/common/src/mmr.rs

```diff
@@ -0,0 +1,152 @@
+		pub BeefyNextAuthoritiesRoot get(fn beefy_next_authorities_root): MerkleRootOf<T>;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Config> for enum Call where origin: <T as frame_system::Config>::Origin {
+	}
+}
+
+impl<T: Config> LeafDataProvider for Module<T> where
+	MerkleRootOf<T>: From<H256>,
+{
+	type LeafData = MmrLeaf<
+		<T as frame_system::Config>::BlockNumber,
+		<T as frame_system::Config>::Hash,
+		MerkleRootOf<T>,
+	>;
+
+	fn leaf_data() -> Self::LeafData {
+		MmrLeaf {
+			parent_number_and_hash: frame_system::Module::<T>::leaf_data(),
+			parachain_heads: Module::<T>::parachain_heads_merkle_root(),
+			beefy_next_authority_set: (
+				Module::<T>::beefy_next_authority_set_id(),
+				Module::<T>::beefy_next_authority_set_merkle_root()
+			),
+		}
+	}
+}
+
+impl<T: Config> Module<T> where
+	MerkleRootOf<T>: From<H256>,
+{
+	/// Returns latest root hash of a merkle tree constructed from all registered parachain headers.
+	///
+	/// NOTE this does not include parathreads - only parachains are part of the merkle tree.
+	///
+	/// TODO [ToDr] describe merkle tree construction.
+	///
+	/// NOTE This is an initial and inefficient implementation, which re-constructs
+	/// the merkle tree every block. Instead we should update the merkle root in [Self::on_initialize]
+	/// call of this pallet and update the merkle tree efficiently (use on-chain storage to persist inner nodes).
+	fn parachain_heads_merkle_root() -> MerkleRootOf<T> {
+		let para_heads = paras::Module::<T>::parachains()
+			.into_iter()
+			.map(paras::Module::<T>::para_head)
+			.map(|maybe_para_head| maybe_para_head.encode())
+			.collect::<Vec<_>>();
+
+		sp_io::trie::keccak_256_ordered_root(para_heads).into()
+	}
+
+	/// Returns a merkle root of a tree constructed from secp256k1 public keys of current BEEFY authority set.
+	///
+	/// NOTE This is an initial and inefficient implementation, which re-constructs
+	/// the merkle tree every block. Instead we should update the merkle root in [on_new_session]
+	/// callback, cause we know it will only change every session - in future it should be optimized
+	/// to change every era instead.
+	fn beefy_next_authority_set_merkle_root() -> MerkleRootOf<T> {
+		Self::beefy_next_authorities_root()
+	}
+
+	fn beefy_next_authority_set_id() -> ValidatorSetId {
+		pallet_beefy::Module::<T>::validator_set_id() + 1
+	}
+
+	fn update_beefy_next_authorities() {
+		let beefy_public_keys = pallet_beefy::Module::<T>::next_authorities()
+			.into_iter()
+			.map(|authority_id| authority_id.encode())
+			.collect::<Vec<_>>();
+		let merkle_root: MerkleRootOf<T> = sp_io::trie
+			::keccak_256_ordered_root(beefy_public_keys).into();
+		BeefyNextAuthoritiesRoot::<T>::put(merkle_root)
+	}
+}
+
+impl<T: Config> sp_runtime::BoundToRuntimeAppPublic for Module<T> {
+	type Public = <T as pallet_beefy::Config>::AuthorityId;
+}
+
+impl<T: Config> frame_support::traits::OneSessionHandler<T::AccountId> for Module<T> where
```

Comment:
This `OneSessionHandler` would have to run after the `pallet_beefy::OneSessionHandler` right? Otherwise `pallet_beefy` storage will not have been updated. Is that guranteed?

Reply:
Good questions. AFAIR hooks are being run in the same order as pallets are listed in `construct_runtime`. I've updated the code now, so `MmrLeaf` pallet is last.